common/src/reserved.rs

```diff
@@ -59,18 +62,72 @@ impl ReservedState {
                     .or_insert(member.consensus_voting_power);
             }
         }
-        Ok(governance_set
-            .iter()
-            .map(|(name, voting_power)| (self.query_public_key(name).unwrap(), *voting_power))
-            .collect())
+        let mut result = Vec::new();
+        for (name, voting_power) in governance_set {
+            let public_key = self.query_public_key(&name).ok_or_else(|| {
+                format!("The public key of {name} is not found in the genesis info.")
+            })?;
+            result.push((public_key, voting_power));
+        }
+        Ok(result)
     }
 
-    pub fn apply_delegate(&mut self, _tx: &TxDelegate) -> Result<Self, String> {
-        unimplemented!()
+    pub fn apply_delegate(&mut self, tx: &TxDelegate) -> Result<Self, String> {
+        let data = DelegationTransactionData {
+            delegator: tx.delegator.clone(),
+            delegatee: tx.delegatee.clone(),
+            governance: tx.governance,
+            timestamp: tx.timestamp as BlockHeight,
+        };
+        match tx.proof.verify(&data) {
+            Ok(_) => true,
+            Err(_) => return Err("proof verification failed".to_string()),
+        };
+        let delegator_name = match self.query_name(&tx.delegator.clone()) {
+            Some(name) => name,
+            None => return Result::Err("delegator does not exist by name".to_string()),
+        };
+        let delegatee_name = match self.query_name(&tx.delegatee.clone()) {
+            Some(name) => name,
+            None => return Result::Err("delegatee does not exist by name".to_string()),
+        };
+        for delegator in &mut self.members {
+            if delegator.name == delegator_name {
+                if tx.governance {
+                    delegator.governance_delegatee = Option::from(delegatee_name.clone());
+                    delegator.consensus_delegatee = Option::from(delegatee_name.clone());
+                } else {
+                    delegator.consensus_delegatee = Option::from(delegatee_name.clone());
+                }
+            }
+        }
+        Ok(self.clone())
     }
 
-    pub fn apply_undelegate(&mut self, _tx: &TxUndelegate) -> Result<Self, String> {
-        unimplemented!()
+    pub fn apply_undelegate(&mut self, tx: &TxUndelegate) -> Result<Self, String> {
+        let data = UndelegationTransactionData {
+            delegator: tx.delegator.clone(),
+            timestamp: tx.timestamp as BlockHeight,
+        };
+        match tx.proof.verify(&data) {
+            Ok(_) => true,
+            Err(_) => return Err("proof verification failed".to_string()),
```

Comment:
```suggestion
            Err(_) => return Err("undelegation proof verification failed".to_string()),
```
For the same reason as above.

Reply:
okay. would change it.